xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,499 @@
+    ManagedChannel channel = ManagedChannelBuilder.forTarget(serverUri).build();
+    startDiscoveryRpc(channel);
+  }
+
+  @Override
+  void shutdown() {
+    channel.shutdown();
+    channel = null;
+    shutdownDiscoveryRpc();
+  }
+
+  /**
+   * Creates a new RPC stream on the given channel for xDS protocol communication
+   * and starts with a LDS request.
+   */
+  @VisibleForTesting
+  void startDiscoveryRpc(ManagedChannel channel) {
+    checkState(this.channel == null, "previous channel has not been cleared yet");
+    checkState(adsStream == null, "previous adsStream has not been cleared yet");
+    this.channel = checkNotNull(channel, "channel");
+    AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub =
+        AggregatedDiscoveryServiceGrpc.newStub(channel);
+    adsStream = new AdsStream(stub);
+    adsStream.start();
+    stopwatch.reset().start();
+    adsStream.sendLdsRequest(targetName);
+  }
+
+  @VisibleForTesting
+  void shutdownDiscoveryRpc() {
+    if (adsStream != null) {
+      adsStream.close(Status.CANCELLED.withDescription("shutdown").asException());
+    }
+    if (rpcRetryTimer != null) {
+      rpcRetryTimer.cancel();
+    }
+  }
+
+  /**
+   * Handles LDS response to find the HttpConnectionManager message for the requested resource name.
+   * Sends an ACK request to the management server if the HttpConnectionManager for the requested
+   * resource name is found, which may either contain the RouteConfiguration directly in-line or
+   * contain the configuration to send RDS requests for dynamic resolution. Otherwise, sends back
+   * a NACK request. Proceed with the resolved RouteConfiguration (if exists) to find the
+   * VirtualHost configuration for the "xds:" URI (with the port, if any, stripped off).
+   */
+  private void handleLdsResponse(DiscoveryResponse ldsResponse) {
+    logger.log(Level.FINE, "Received a LDS response: {0}", ldsResponse);
+    adsStream.ldsRespNonce = ldsResponse.getNonce();
+    List<Any> respResources = ldsResponse.getResourcesList();
+    HttpConnectionManager connManager = null;
+    try {
+      for (com.google.protobuf.Any res : respResources) {
+        Listener listener = res.unpack(Listener.class);
+        if (listener.getName().equals(targetName)) {
+          HttpConnectionManager httpConnManager =
+              listener.getApiListener().getApiListener().unpack(HttpConnectionManager.class);
+          if (httpConnManager.hasRouteConfig() || httpConnManager.hasRds()) {
+            connManager = httpConnManager;
+          }
+          // Ignore remaining listeners once the requested one is found.
+          break;
+        }
+      }
+    } catch (InvalidProtocolBufferException e) {
+      configWatcher.onError(Status.fromThrowable(e).augmentDescription("Invalid LDS response"));
+    }
+    if (connManager != null) {
+      // Accept this listener update.
+      adsStream.ldsVersion = ldsResponse.getVersionInfo();
+    }
+    // Send ACK/NACK request.
```

Comment:
You are sending ACK here (and not NACK), may be the comment can reflect that. Also we can delay doing this until after the following code has executed (which might result in sending NACK)

Reply:
Whether a request is an ACK or NACK depends on `version_info`. At this point, `version_info` (aka, `adsStream.ldsVersion`) can be updated to this response's `version_info` (ACK) or still be the previous one (NACK). So This request is either ACK or NACK.

Regarding whether we should send ACK/NACK after processing `HttpConnectionManager` message, it's vague in the design. When the `HttpConnectionManager` message contains the configuration to do RDS, it will send an RDS request. But we need to ACK the LDS response first. 

I am following the condition for ACK/NACK: as long as the `DiscoveryResponse` contains the requested resource, it should be an ACK. Regardless if inner messages contain valid information for the resource. For example, if our LDS request asks for resource `foo.googleapis.com`, an LDS response containing `foo.googleapis.com` as one of its resources should be ACKed, even if its inner message may be broken. This is kind of a "breadth-first" processing. Not sure if this is correct. But if we want to do a "depth-first" processing, we need to go through the inner message twice: first populate inner message to see if it contains valid information to proceed, send ACK/NACK for current response, then proceed with the inner message (which may trigger next steps of request/response). Can you clarify the correct behavior? @markdroth

Reply:
> ...
> I am following the condition for ACK/NACK: as long as the `DiscoveryResponse` contains the requested resource, it should be an ACK. Regardless if inner messages contain valid information for the resource. For example, if our LDS request asks for resource `foo.googleapis.com`, an LDS response containing `foo.googleapis.com` as one of its resources should be ACKed, even if its inner message may be broken. 

That's not what I understood from https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack-nack-and-versioning . Specifically:

"Envoy will send a new request on the stream, specifying the last version **successfully applied** ... If the **update was successfully applied**,... 
... If Envoy had instead rejected configuration update X...."

The above wording suggests ACK is to be sent only after "successfully applying" the configuration and not just matching the expected resource names. But happy to be corrected...

Reply:
That sounds more solid. I updated the implementation. The client will wait to send ACK/NACK until it knows the response is useful or not. By "useful", it means the response either contains the cluster name information or it can leads to a valid next step request (e.g., LDS response configuration to do RDS). This is done by preparing an ACK request when receiving a response, and then send it out when we are about to proceed to next step or change it to an NACK and send out when we are not able to proceed.